### ISSUE TITLE
[Spark functions] mismatch results in 'In' expression

### ISSUE THREAD

**Issue by bikramSingh91** [facebookincubator/velox#3995]
### Description

mismatch results in 'In' expression

### Error Reproduction

[velox_expressionVerifier_s2fRDO.zip](https://github.com/facebookincubator/velox/files/10690589/velox_expressionVerifier_s2fRDO.zip)

### Relevant logs

```bash
libc++abi: terminating with uncaught exception of type facebook::velox::VeloxRuntimeError: Exception: VeloxRuntimeError
Error Source: RUNTIME
Error Code: INVALID_STATE
Reason: Different results at idx '0': 'null' vs. '0'
Retriable: False
Expression: left->equalValueAt(right.get(), i, i)
Function: compareVectors
File: /Users/bikramjeet/dev/velox/velox/expression/tests/ExpressionVerifier.cpp
Line: 66
Stack trace:
```

**Comment by mbasmanova** [comment]
@rui-mo Rui, would Gluten team help fix bugs in Spark functions found by running the Fuzzer.

Here are some recent issues: https://github.com/issues?q=is%3Aopen%20is%3Aissue%20author%3AbikramSingh91%20archived%3Afalse%20Spark%20Functions

**Comment by rui-mo** [comment]
> @rui-mo Rui, would Gluten team help fix bugs in Spark functions found by running the Fuzzer.
> 
> Here are some recent issues: https://github.com/issues?q=is%3Aopen%20is%3Aissue%20author%3AbikramSingh91%20archived%3Afalse%20Spark%20Functions

Yes, as @FelixYBW replied in the email.

**Comment by mbasmanova** [comment]
Thank you, Rui and Binwei.

**Comment by JkSelf** [comment]
@bikramSingh91
I tried to reproduce this problem based on the data you provided, but it didn't work. The following is my steps.  And the code commit is e4a2c99bbf8c61b4cb3edabd5ce1ac25509f0652. Please correct me if wrong operators. 
```
make debug
velox/expression/tests/velox_expression_runner_test --sql_path ~/velox_expressionVerifier_s2fRDO/sql --input_path ~/velox_expressionVerifier_s2fRDO/input_vector --complex_constant_path ~/velox_expressionVerifier_s2fRDO/complex_constants
Result: ROW<_col0:BOOLEAN>
null
null
null
null
true
null
null
null
null
null
```

**Comment by bikramSingh91** [comment]
> @bikramSingh91 I tried to reproduce this problem based on the data you provided, but it didn't work. The following is my steps. And the code commit is [e4a2c99](https://github.com/facebookincubator/velox/commit/e4a2c99bbf8c61b4cb3edabd5ce1ac25509f0652). Please correct me if wrong operators.
> 
> ```
> make debug
> velox/expression/tests/velox_expression_runner_test --sql_path ~/velox_expressionVerifier_s2fRDO/sql --input_path ~/velox_expressionVerifier_s2fRDO/input_vector --complex_constant_path ~/velox_expressionVerifier_s2fRDO/complex_constants
> Result: ROW<_col0:BOOLEAN>
> null
> null
> null
> null
> true
> null
> null
> null
> null
> null
> ```

@JkSelf This is because currently the expressionRunner has presto functions registered so its using the preso equivalent function which works correctly. See https://github.com/facebookincubator/velox/issues/3996 for more details.

To workaround this, I have created a diff that you can apply which makes the required changes to run with spark functions. You can then try to run the command on velox_expression_runner_test again.

diff: 
[sparkExpressionRunner.diff.zip](https://github.com/facebookincubator/velox/files/10808948/sparkExpressionRunner.diff.zip)
(had to compress the diff file because github wont allow me to upload it otherwise)

**Comment by zhli1142015** [comment]
Hello,
I observed this error also in my local test for https://github.com/facebookincubator/velox/pull/7805#pullrequestreview-1757408767. 
`
./spark_expression_fuzzer_test --duration_sec 600 --enable_variadic_signatures --lazy_vector_generation_ratio 0.2 --velox_fuzzer_enable_column_reuse --velox_fuzzer_enable_expression_reuse --max_expression_trees_per_step 1 --retry_with_try --logtostderr=1 --minloglevel=0 --only in,array_sort,array,if --special_forms switch,and,or --velox_fuzzer_enable_complex_types
`
From my investigation the result difference between common and simplified evals is because of constant folding. In common eval, the second argument would be converted to constant vectors, but not for simplified eval. This caused different values of [`hasNull_`](https://github.com/facebookincubator/velox/blob/main/velox/functions/sparksql/In.cpp#L99), then different results.

Please let me know if I can do the fix for this item?
cc @rui-mo and @mbasmanova .
Thanks.

**Comment by zhli1142015** [comment]
> Hello, I observed this error also in my local test for [#7805 (review)](https://github.com/facebookincubator/velox/pull/7805#pullrequestreview-1757408767). `./spark_expression_fuzzer_test --duration_sec 600 --enable_variadic_signatures --lazy_vector_generation_ratio 0.2 --velox_fuzzer_enable_column_reuse --velox_fuzzer_enable_expression_reuse --max_expression_trees_per_step 1 --retry_with_try --logtostderr=1 --minloglevel=0 --only in,array_sort,array,if --special_forms switch,and,or --velox_fuzzer_enable_complex_types` From my investigation the result difference between common and simplified evals is because of constant folding. In common eval, the second argument would be converted to constant vectors, but not for simplified eval. This caused different values of [`hasNull_`](https://github.com/facebookincubator/velox/blob/main/velox/functions/sparksql/In.cpp#L99), then different results.
> 
> Please let me know if I can do the fix for this item? cc @rui-mo and @mbasmanova . Thanks.

From the implmention of Spark [In](https://github.com/facebookincubator/velox/blob/main/velox/functions/sparksql/In.cpp#L71) function, the second argument should be marked as the constant argument. Then fuzzer would only genernate constant parameters for it.
Looks [constatn argument support](https://github.com/facebookincubator/velox/issues/4140) in the function signatures is added. But constant argument is not supported in the simple function API. I think we can fix this issue by adding constant argument support to the simple function API.
1. define static const function to tell which argument is constant in "In" function
`static constexpr bool is_constant_arg(int32_t position)`
2. check if argument is constant, we call `constantArgumentType` for it.
```
    int32_t position = 0;
    for (const auto& arg : analysis.argsTypes) {
      if (udf_is_constant_arg<Fun>()(position)) {
        builder.constantArgumentType(arg);
      } else {
        builder.argumentType(arg);
      }
      position++;
    }
```
@rui-mo and @mbasmanova , please let me know if this looks a correct fix for you.
Thanks.

**Comment by mbasmanova** [comment]
>  I think we can fix this issue by adding constant argument support to the simple function API.

That's right. Let's create GitHub issue to discuss how exactly this can be done. CC: @laithsakka @kagamiori
